src/server/saved_objects/migrations/core/migrate_raw_docs.ts

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * This file provides logic for migrating raw documents.
+ */
+
+import { RawDoc } from './call_cluster';
+import { SavedObjectDoc, TransformFn } from './document_migrator';
+
+/**
+ * Applies the specified migration function to every saved object document in the list
+ * of raw docs. Any raw docs that are not valid saved objects will simply be passed through.
+ *
+ * @param {TransformFn} migrateDoc
+ * @param {RawDoc[]} rawDocs
+ * @returns {RawDoc[]}
+ */
+export function migrateRawDocs(migrateDoc: TransformFn, rawDocs: RawDoc[]): RawDoc[] {
+  return rawDocs.map(
+    raw => (isRawSavedObject(raw) ? savedObjectToRaw(migrateDoc(rawToSavedObject(raw))) : raw)
+  );
+}
+
+/**
+ * Determines whether or not the raw document can be converted to a saved object.
+ */
+function isRawSavedObject(rawDoc: RawDoc) {
+  const { type } = rawDoc._source;
+  return type && rawDoc._id.startsWith(type) && rawDoc._source.hasOwnProperty(type);
```

Comment:
@elastic/kibana-security will this effect spaces, or did we decide to still start with the type?

Reply:
If I understand what's happening here, then I do think this will impact Spaces. We're planning on introducing the concept of "namespaces" into the OSS Saved Objects Repository, which will prefix the document id with a namespace.

We'll have a PR against master which implements this soon, but you can see the original PR here: https://github.com/elastic/kibana/pull/22357/files 

To distill it down, this is how we are planning on generating document ids: https://github.com/kobelb/kibana/blob/95ef4b02bd525251354846b259df46e4173e8d3c/src/server/saved_objects/service/lib/repository.js#L479

If the migration used the `repository` to access saved objects, then it might be isolated from this, but I'm guessing it isn't able to take advantage of the repository, and has to query using `callCluster` directly?

Reply:
Even if we aren't able to use the repository directly from the migrations, it seems like the `rawToSavedObject` and `savedObjectToRaw` could be generalized below for use within the Repository and the migrations so that we have a single source of this logic.

Reply:
Chatted with @kobelb - can we change this check to validate the property `type` also has a corresponding `type` property? This is more what defines a saved object and will eliminate the discrepancy with Spaces taking over the ID.

Reply:
Yeah. We're operating on the raw index, for a variety of reasons. I agree w/ @kobelb that this logic needs to be isolated in a way that it can be overridden by x-pack / whatever, and reused anywhere that this kind of thing needs to be done.